File: src/ruby/selection_sort.rb
# frozen_string_literal: true

# Sort an array using the SelectionSort algorithm
class SelectionSort
  attr_reader :array_sorted

  def initialize
    @array_sorted = []
  end

  def init(array)
    selection_sort(array, array.length - 1)
  end

  private

  def selection_sort(array, size)
    size.times do |i|
      index_min = i

      (i + 1).upto(size) do |j|
        index_min = j if array[j] < array[index_min]
      end
      array[i], array[index_min] = array[index_min], array[i] if index_min != i
    end
    @array_sorted = array
  end
end

# test
s_s = SelectionSort.new
s_s.init([1, 4, 10, 2, 3, 32, 0])
p s_s.array_sorted

File: src/go/selection_sort.go
package main

import "fmt"

func SelectionSort(slice []int) {
	for index1 := 0; index1 < len(slice)-1; index1++ {
		indexMin := index1

		for index2 := index1 + 1; index2 < len(slice); index2++ {
			if slice[index2] < slice[indexMin] {
				indexMin = index2
			}
		}

		if index1 != indexMin {
			slice[index1], slice[indexMin] = slice[indexMin], slice[index1]
		}
	}
}

func main() {
  slice := []int{1, 3, 2, 9, 7, 8, 5, 6, 4}
	fmt.Println("Slice:", slice)
	SelectionSort(slice)
	fmt.Println("SelectionSort:", slice)
}
